File: Epsitec.Cresus/Cresus.WebCore.Server/Owin/Hubs/EntityBagHub.cs
//	Copyright © 2013, EPSITEC SA, CH-1400 Yverdon-les-Bains, Switzerland
//	Author: Samuel LOUP, Maintainer: Samuel LOUP

using Microsoft.AspNet.SignalR;

using System.Threading.Tasks;

namespace Epsitec.Cresus.WebCore.Server.Owin.Hubs
{
	public class EntityBagHub : Hub
	{

		/*public void AddToBag(string connectionId, string title, string summary, string entityId)
		{
			Clients.Client (connectionId).AddToBag (title, summary, entityId);
		}

		public void RemoveFromBag(string connectionId, string title, string summary, string entityId)
		{
			Clients.Client (connectionId).RemoveFromBag (title, summary, entityId);
		}*/


		public override Task OnDisconnected()
		{
			var backendClient = EntityBagClient.Instance;


			Clients.Client (backendClient.GetConnectionId ()).FlushConnectionId (Context.ConnectionId);

			return base.OnDisconnected ();
		}

		public override Task OnReconnected()
		{
			var backendClient = EntityBagClient.Instance;

			Clients.Client (backendClient.GetConnectionId ()).SetUserConnectionId (Clients.Caller.userName, Context.ConnectionId);
			return base.OnReconnected ();
		}

		public void SetupUserConnection()
		{
			var backendClient = EntityBagClient.Instance;
			Clients.Client (backendClient.GetConnectionId ()).SetUserConnectionId (Clients.Caller.userName, Clients.Caller.connectionId);
		}

		public void RemoveFromMyBag(string entityId)
		{
			var backendClient = EntityBagClient.Instance;
			Clients.Client (backendClient.GetConnectionId ()).RemoveFromMyBag (Clients.Caller.userName, entityId);
		}

		public void AddToMyBag(string title, string customSummary,string entityId)
		{
			var backendClient = EntityBagClient.Instance;
			Clients.Client (backendClient.GetConnectionId ()).AddToMyBag (Clients.Caller.userName, title, customSummary, entityId);
		}
	}
}

File: Epsitec.Cresus/Cresus.WebCore.Client/WebCore/js/hub/EntityBagHub.js
function EntityBagHub() {
  this.hub = $.connection.entityBagHub;
  var entityBagClient = null;

  //Initialize
  this.init = function(con,username, client) {   
    this.hub = con.entityBagHub;
    this.hub.state.connectionId = this.hub.connection.id;
    this.hub.state.userName = username;
    entityBagClient = client;
    this.hub.server.setupUserConnection();
  };


  //Entry points for calling hub
  this.RemoveFromMyBag = function(entityId) {
    this.hub.server.removeFromMyBag(entityId);

  };

  this.AddToMyBag = function(title,summary,entityId) {
    this.hub.server.addToMyBag(title,summary,entityId);
  };

  //Entry points for hub call
  this.hub.client.AddToBag = function(title, summary, entityId) {
    var app = Epsitec.Cresus.Core.getApplication();
    var entity = {
          summary: summary,
          entityType: title,
          id: entityId
        };

    app.addEntityToClientBag(entity);
  };

  this.hub.client.RemoveFromBag = function(entityId) {
    var app = Epsitec.Cresus.Core.getApplication();
    var entity = {
          id: entityId
        };

    app.removeEntityFromClientBag(entity);
  };

  this.hub.client.SetLoading = function(state) {
    var app = Epsitec.Cresus.Core.getApplication();
    app.viewport.setLoading(state);
  };
}
